### PR TITLE
Adds Maintenance Policy to Memorystore module

### LINKED ISSUE #66: Maintenance Policy for Redis
Hi, going through the code I am unable to find parameters to define maintenance policy or schedule for redis. Is the feature supported, if yes, can you please redirect me to some documentation related to it.
Thanks in advance.

### PR DESCRIPTION
**Pull request by jamiedick**
Closes #66

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: variables.tf [hunk 1]
```diff
@@ -125,3 +125,25 @@ variable "transit_encryption_mode" {
...
+  })
+  default = {
+    day = "SATURDAY"
+    start_time = {
+      hours   = 4
+      minutes = 30
+      seconds = 0
+      nanos   = 0
+    }
+  }
```
Comment (bharathkkb) [inline variables.tf:148]: Let's set the default to empty to maintain backward compatibility and use api default.

**Review by bharathkkb — CHANGES_REQUESTED** [review]
Thanks for the PR @jamiedick

**Comment by bharathkkb** [comment]
Looks like there is an inflight PR https://github.com/terraform-google-modules/terraform-google-memorystore/pull/79
I will ping the author and see if they want to finish that otherwise happy to proceed with this.

File: variables.tf (hunk 1 above)
Comment (jamiedick) [reply variables.tf:148]: @bharathkkb maybe I'm missing something, but to define the variable type as such:
```
  type = object({
    day = string
    start_time = object({
      hours   = number
      minutes = number
      seconds = number
      nanos   = number
    })
  })
```
will require the default value to be set with something other than: `default: {}`

**Comment by jamiedick** [comment]
@bharathkkb both requests updated - thanks!

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/README.md /tmp/tmp.Y5LiRo4o4e/generate_docs/workspace/README.md
39c39
< | maintenance\_policy | The maintenance policy for an instance. | <pre>object({<br>    day = string<br>    start_time = object({<br>      hours   = number<br>      minutes = number<br>      seconds = number<br>      nanos   = number<br>    })<br>  })</pre> | <pre>{<br>  "day": "SATURDAY",<br>  "start_time": {<br>    "hours": 4,<br>    "minutes": 30,<br>    "nanos": 0,<br>    "seconds": 0<br>  }<br>}</pre> | no |
---
> | maintenance\_policy | The maintenance policy for an instance. | <pre>object({<br>    day = string<br>    start_time = object({<br>      hours   = number<br>      minutes = number<br>      seconds = number<br>      nanos   = number<br>    })<br>  })</pre> | `null` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
./main.tf:58:    
Error: Trailing whitespace found in the lines above.
Checking for missing newline at end of file
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
main.tf
--- old/main.tf
+++ new/main.tf
@@ -42,7 +42,7 @@
 
   transit_encryption_mode = var.transit_encryption_mode
 
-  dynamic maintenance_policy {
+  dynamic "maintenance_policy" {
     for_each = var.maintenance_policy != null ? [var.maintenance_policy] : []
     content {
       weekly_maintenance_window {
@@ -55,7 +55,7 @@
         }
       }
     }
-    
+
   }
 }
 
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_whitespace
 Failed whitespace check. More details below. 
 ```bash 
Checking for trailing whitespace
./main.tf:58:    
Error: Trailing whitespace found in the lines above.
Checking for missing newline at end of file
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
main.tf
--- old/main.tf
+++ new/main.tf
@@ -42,7 +42,7 @@
 
   transit_encryption_mode = var.transit_encryption_mode
 
-  dynamic maintenance_policy {
+  dynamic "maintenance_policy" {
     for_each = var.maintenance_policy != null ? [var.maintenance_policy] : []
     content {
       weekly_maintenance_window {
@@ -55,7 +55,7 @@
         }
       }
     }
-    
+
   }
 }
 
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by jamiedick** [comment]
@bharathkkb any other feedback or can we move forward with this PR?

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Review by bharathkkb — COMMENTED** [review]
@jamiedick Lets also bump min version to 4.1.0 https://github.com/terraform-google-modules/terraform-google-memorystore/blob/0c0dd292c75833ca032f7a25e11a2c53ba90e3ea/versions.tf#L23 when this was [introduced in provider](https://github.com/terraform-google-modules/terraform-google-memorystore/blob/0c0dd292c75833ca032f7a25e11a2c53ba90e3ea/versions.tf#L23). 

Since this will be a breaking change, if you have bandwidth it would also be good to add this same block to [memcache](https://github.com/terraform-google-modules/terraform-google-memorystore/blob/master/modules/memcache/main.tf) submodule in the repo and bump that [version to 4.23](https://github.com/hashicorp/terraform-provider-google-beta/blob/main/CHANGELOG.md#4230-june-1-2022).

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate .
Success!
 The configuration is valid.
terraform_validate ./examples/basic
Success!
 The configuration is valid.
terraform_validate ./examples/memcache
╷
│
 
Error: 
Failed to query available provider packages
│
 
│
 
Could not retrieve the list of available versions for provider
│
 
hashicorp/google-beta: no available releases match the given constraints >=
│
 
3.43.0, >= 3.45.0, ~> 3.52, >= 3.53.0, >= 4.23.0, < 5.0.0
╵
╷
│
 
Error: 
Failed to query available provider packages
│
 
│
 
Could not retrieve the list of available versions for provider
│
 
hashicorp/google: no available releases match the given constraints >=
│
 
2.12.0, >= 3.43.0, >= 3.45.0, ~> 3.52, >= 3.53.0, >= 3.83.0, >= 4.23.0, <
│
 
5.0.0
╵
terraform_validate ./examples/minimal
Success!
 The configuration is valid.
terraform_validate ./examples/redis
Success!
 The configuration is valid.
terraform_validate ./modules/memcache
╷
│
 
Error: 
Missing required argument
│
 
│
 
  on main.tf line 44, in resource "google_memcache_instance" "self":
│
 
  44:       weekly_maintenance_window 
{
│
 
│
 
The argument "duration" is required, but no definition was found.
╵
terraform_validate ./test/setup
Success!
 The configuration is valid.
 ```

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
modules/memcache/main.tf
--- old/modules/memcache/main.tf
+++ new/modules/memcache/main.tf
@@ -42,7 +42,7 @@
     for_each = var.maintenance_policy != null ? [var.maintenance_policy] : []
     content {
       weekly_maintenance_window {
-        day = maintenance_policy.value["day"]
+        day      = maintenance_policy.value["day"]
         duration = maintenance_policy.value["duration"]
         start_time {
           hours   = maintenance_policy.value["start_time"]["hours"]
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
modules/memcache/variables.tf
--- old/modules/memcache/variables.tf
+++ new/modules/memcache/variables.tf
@@ -93,7 +93,7 @@
   description = "The maintenance policy for an instance."
   # type = object(any)
   type = object({
-    day = string
+    day      = string
     duration = number
     start_time = object({
       hours   = number
Error: terraform fmt failed with exit code 3
Check the output for diffs and correct using terraform fmt <dir>
 ```

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/memcache/README.md /tmp/tmp.6gknPmZ2BR/generate_docs/workspace/modules/memcache/README.md
15c15
< | maintenance\_policy | The maintenance policy for an instance. | <pre>object({<br>    day = string<br>    duration = number<br>    start_time = object({<br>      hours   = number<br>      minutes = number<br>      seconds = number<br>      nanos   = number<br>    })<br>  })</pre> | `null` | no |
---
> | maintenance\_policy | The maintenance policy for an instance. | <pre>object({<br>    day      = string<br>    duration = number<br>    start_time = object({<br>      hours   = number<br>      minutes = number<br>      seconds = number<br>      nanos   = number<br>    })<br>  })</pre> | `null` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ```

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Comment by jamiedick** [comment]
@bharathkkb makes sense - i pushed those changes! let me know if theres anything else.

File: versions.tf [hunk 2]
```diff
@@ -20,7 +20,7 @@ terraform {
 
     google = {
       source  = "hashicorp/google"
-      version = ">= 3.53, < 5.0"
+      version = ">= 4.1.0, < 5.0"
```
Comment (bharathkkb) [inline versions.tf:23]: Sorry typo, it was in 4.10 https://github.com/hashicorp/terraform-provider-google-beta/blob/main/CHANGELOG.md#4100-february-7-2022
```suggestion
      version = ">= 4.10.0, < 5.0"
```

**Review by bharathkkb — APPROVED** [review]
LGTM! Will merge after tests pass

**Comment by comment-bot-dev** [comment]
@jamiedick
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

**Pull request merged**